tools/cell_census_builder/validate.py

```diff
@@ -278,56 +278,49 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
+    4. the number of present genes in each h5ad dataset == the row sum of the corresponding datasets in the presence
       matrix.
     """
-    error = "{}:{} 'raw' {} mismatch {} vs {}"
-
     assets_path, soma_path, dataset, experiment_specifications = args
     _, unfiltered_ad = next(open_anndata(assets_path, [dataset]))
 
     for eb in experiment_specifications:
+        error = f"{eb.name}:{dataset.dataset_id}" + " 'raw' {} mismatch {} vs {}"
         with open_experiment(soma_path, eb) as exp:
             anndata_cell_filter = make_anndata_cell_filter(eb.anndata_cell_filter_spec)
             ad = anndata_cell_filter(unfiltered_ad, retain_X=True)
-
-            # Extract soma_joinids for cells associated with the dataset in X["raw"].
-            cell_soma_joinids: npt.NDArray[np.int64] = (
-                exp.obs.read(
-                    column_names=["soma_joinid", "dataset_id"], value_filter=f"dataset_id == '{dataset.dataset_id}'"
-                )
-                .concat()
-                .to_pandas()
-                .soma_joinid.to_numpy()
-            )
-
-            raw_sum = 0.0
-            raw_nnz = 0
-            fdpm_gene_count = 0
-            if len(cell_soma_joinids) > 0:
-                assert soma.SparseNDArray.exists(exp.ms[MEASUREMENT_RNA_NAME].X["raw"].uri)
-                raw_nnz = _count_elements(exp.ms[MEASUREMENT_RNA_NAME].X["raw"], cell_soma_joinids)
-                raw_sum = _sum_elements(exp.ms[MEASUREMENT_RNA_NAME].X["raw"], cell_soma_joinids)
-                fdpm_gene_count = _count_elements(
-                    exp.ms["RNA"][FEATURE_DATASET_PRESENCE_MATRIX_NAME], [dataset.soma_joinid]
-                )
-
+            ad_genes = [*set(ad.var.index.array)]
             if ad.raw is None:
                 ad_x = ad.X
             else:
                 ad_x = ad.raw.X
-            h5ad_sum = sum(ad_x.data)  # calculate the actual sum of the X matrix values
-            assert raw_sum == h5ad_sum, error.format(eb.name, dataset.dataset_id, "sum", raw_sum, h5ad_sum)
-            h5ad_nnz = _count_nonzero(ad_x.data)
-            assert raw_nnz == h5ad_nnz, error.format(eb.name, dataset.dataset_id, "nnz", raw_nnz, h5ad_nnz)
+
+            with exp.axis_query(
+                measurement_name="RNA",
+                obs_query=soma.AxisQuery(value_filter=f"dataset_id == '{dataset.dataset_id}'"),
+                var_query=soma.AxisQuery(value_filter=f"feature_id in {ad_genes}"),
+            ) as query:
+                fdpm_gene_count = 0
+                raw_coo = None
+                if query.n_obs > 0:
+                    assert soma.SparseNDArray.exists(exp.ms[MEASUREMENT_RNA_NAME].X["raw"].uri)
+                    fdpm_gene_count = _count_elements(
+                        exp.ms["RNA"][FEATURE_DATASET_PRESENCE_MATRIX_NAME], [dataset.soma_joinid]
+                    )
+                    raw_coo = query.to_anndata("raw").X.tocoo()  # TODO: slow, find a faster method.
```

Comment:
@bkmartinjr is there a better way of converting this?

Reply:
Maybe try to avoid the `to_anndata` conversion: `query.X('raw').coos().concat().to_scipy()`

Reply:
This gives the wrong shape I run this with `test_base_builder_creation`. Converting to anndata has been the only way I've found so far to get the h5ad and cell_census to have the same shape.